java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/coverage/CoverageHandler.java

```diff
@@ -0,0 +1,182 @@
+            IPath outputLocation = entry.getOutputLocation();
+            if (outputLocation == null) {
+                outputLocation = javaProject.getOutputLocation();
+            }
+            final IPath outputRelativePath = outputLocation.makeRelativeTo(javaProject.getProject().getFullPath());
+            outputToSourcePaths.computeIfAbsent(outputRelativePath, k -> new LinkedList<>()).add(sourceRelativePath);
+        }
+        return outputToSourcePaths;
+    }
+
+    /**
+     * Infer the source file for the given {@link ISourceFileCoverage}. If no file found, return <code>null</code>.
+     */
+    private File getSourceFile(List<IPath> sourceRoots, ISourceFileCoverage sourceFileCoverage) {
+        final String packagePath = sourceFileCoverage.getPackageName().replace(".", "/");
+        final IPath sourceRelativePath = new org.eclipse.core.runtime.Path(packagePath)
+                .append(sourceFileCoverage.getName());
+        for (final IPath sourceRoot : sourceRoots) {
+            final IPath relativePath = sourceRoot.append(sourceRelativePath);
+            final File sourceFile = getFileForFs(javaProject, relativePath);
+            if (sourceFile.exists()) {
+                return sourceFile;
+            }
+        }
+        return null;
+    }
+
+    private static File getFileForFs(IJavaProject javaProject, IPath path) {
+        return javaProject.getProject().getLocation().append(path).toFile();
+    }
+
+    private List<LineCoverage> getLineCoverages(final ISourceFileCoverage sourceFileCoverage) {
+        final List<LineCoverage> lineCoverages = new LinkedList<>();
+        final int last = sourceFileCoverage.getLastLine();
+        for (int nr = sourceFileCoverage.getFirstLine(); nr <= last; nr++) {
+            final ILine line = sourceFileCoverage.getLine(nr);
+            if (line.getStatus() != ICounter.EMPTY) {
+                final List<BranchCoverage> branchCoverages = new LinkedList<>();
+                for (int i = 0; i < line.getBranchCounter().getTotalCount(); i++) {
+                    branchCoverages.add(new BranchCoverage(
+                        i < line.getBranchCounter().getCoveredCount() ? 1 : 0));
+                }
+                lineCoverages.add(new LineCoverage(
+                    nr,
+                    line.getInstructionCounter().getCoveredCount(),
+                    branchCoverages
+                ));
+            }
+        }
+        return lineCoverages;
+    }
+
+    private List<MethodCoverage> getMethodCoverages(final CoverageBuilder coverageBuilder,
+            final ISourceFileCoverage sourceFileCoverage) {
+        final List<MethodCoverage> methodCoverages = new LinkedList<>();
+        for (final IClassCoverage classCoverage : coverageBuilder.getClasses()) {
+            if (classCoverage.getSourceFileName().equals(sourceFileCoverage.getName())) {
```

Comment:
should consider the package information as well.

Reply:
since you have refactored the code, and the classCoverages is already belonged to the sourceFileCoverage, this `if` comparison is not necessary any more and can be removed.